backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java

```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
     permissionRepository.delete(permission);
   }
 
+  @Override
+  public PermissionResponse grantPermissionForFile(PermissionGrantRequest permissionGrantRequest) {
+    UserObjectAccessEntity permission = new UserObjectAccessEntity();
+
+    UserEntity currentUser =
+        userRepository
+            .findById(currentUserService.getLoggedUser().getUuid())
+            .orElseThrow(UserNotFoundException::new);
+
+    UserEntity targetUser =
+        userRepository
+            .findById(permissionGrantRequest.getGrantedTo())
+            .orElseThrow(UserNotFoundException::new);
+
+    AccessTypeEntity accessType =
+        accessTypeRepository
+            .findByName(permissionGrantRequest.getAccess())
+            .orElseThrow(AccessTypeNotFoundException::new);
+
+    FileMetadataEntity file =
+        fileMetadataRepository
+            .findById(permissionGrantRequest.getObjectId())
+            .orElseThrow(BucketNotFoundException::new);
+
+    if (file.isDeleted()) {
+      throw new FileAlreadyDeletedException();
+    }
+
+    permission.setGrantedBy(currentUser);
+    permission.setGrantedTo(targetUser);
+    permission.setAccess(accessType);
+    permission.setObject(file);
+    permission.setId(
+        new UserObjectAccessKey(permission.getGrantedTo().getId(), permission.getObject().getId()));
+
+    currentUserHasPermissionForBucket(permissionGrantRequest.getObjectId(), AccessType.OWNER);
+
+    if (permissionRepository.existsById(
+        new UserObjectAccessKey(targetUser.getId(), file.getId()))) {
+      throw new PermissionAlreadyExistsException();
+    }
+
+    UserObjectAccessEntity savedPermission = permissionRepository.save(permission);
+
+    grantViewPermissionToAllFoldersUpstream(file.getParentFolder(), currentUser, targetUser);
+
+    permissionGrantRequest.setObjectId(file.getBucket().getId());
+    permissionGrantRequest.setAccess("view");
+    grantPermissionForBucket(permissionGrantRequest);
```

Comment:
```
  grantViewPermissionToAllFoldersUpstream(file.getParentFolder(), currentUser, targetUser);

  permissionGrantRequest.setObjectId(file.getBucket().getId());
  permissionGrantRequest.setAccess("view");
  grantPermissionForBucket(permissionGrantRequest);
```
this chunk could probably be a method, grant view permission to upstream (objects)

Reply:
also, won't `grantPermissionForBucket()` throw an error if the user already has permission for the bucket?

Reply:
1) Yeah imma extract that
2) Wanted to reuse the method, forgot about that. I think it's easiest to surround it with a try catch